src/transformers/trainer.py

```diff
@@ -1036,6 +1038,14 @@ def train(
         # add remaining tr_loss
         self._total_loss_scalar += tr_loss.item()
 
+        if self.deepspeed:
+            # free up any memory that might be useful for eval
+            self.deepspeed = None
+            self.optimizer = None
+            self.lr_scheduler = None
+            self.model_wrapped = None
```

Comment:
I'm not super fond of having this done automatically. It should be in a `free_memory` method of the `Trainer` (or a name you like better) that is explicitly called by the user between training and evaluation IMO.
This is also useful beyond deepspeed.

Reply:
These are 2 different things.

1. in the case of DeepSpeed this is a clean cut, we explicitly init all these at the beginning of `train`:

https://github.com/pytorch/pytorch/blob/c791a304847ba5216b53acc1225610734a0c8cde/torch/distributed/pipeline/sync/skip/skippable.py#L239

so this PR explicitly cleans these up explicitly at the end of `train` - this is completely opaque. a user had no way to init those explicitly and thus has no need to do anything special.

2. wrt generic case, the story is different because the user may supply her own optimizer/lr_scheduler and in such case, yes, they need to have control over whether to clean up or not.

As you pointed out this would be useful to the user, but it's a different situation, so let's solve it separately?

Reply:
though I do think I need to fix this:
```
            self.model_wrapped = None
```
to restore this to `self.model`

Reply:
Ah yes, it's true that in this case they are instantiated inside the `train` method so this makes sense.